staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -41,10 +41,27 @@ import (
 	"k8s.io/apiserver/pkg/admission/plugin/webhook/generic"
 	webhookrequest "k8s.io/apiserver/pkg/admission/plugin/webhook/request"
 	"k8s.io/apiserver/pkg/admission/plugin/webhook/util"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
 	webhookutil "k8s.io/apiserver/pkg/util/webhook"
 	utiltrace "k8s.io/utils/trace"
 )
 
+const (
+	// PatchAuditAnnotationPrefix is a prefix for persisting webhook patch in audit annotation.
+	// Audit handler decides whether annotation with this prefix should be logged based on audit level.
+	// Since mutating webhook patches the request body, audit level must be greater or equal to Request
+	// for the annotation to be logged
+	PatchAuditAnnotationPrefix = "patch.webhook.admission.k8s.io/"
```

Comment:
I did some research and there are fundamental problems with how the annotation keys should look like. Sorry I couldn't make the following shorter.

The goal is to uniquely capture a webhook invocation event, and the json patch being applied in audit annotations. 

**Background**:

In Kubernetes we have the restriction that an annotation key must be a [qualified name](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/apiserver/pkg/admission/attributes.go#L191) in format of:
- `{optional DNS subdomain prefix with no more than 253 characters}/{a qualified name that consists of alphanumeric characters, '-', '_' or '.', with no more than 63 characters}`

and the semantics is:
- `{plugin}.{org}/{key}`

**This PR**:

The original idea was to use `webhook.admission.k8s.io` as `{org}` and webhook name as `{key}`, so that we know there will be no key conflict against [custom annotations](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L198-L203), and it's clear from the prefix that the annotation is added by the apiserver (admission webhook org). 

However, [webhook names](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/pkg/apis/admissionregistration/validation/validation.go#L261) are allowed to be DNS subdomain with no more than 253 characters, which means webhook name > 63 characters won't get logged. 

**New proposal:**

I propose we use the following semantics for the annotation keys:
- `{webhook name}/k8s_{patch|mutation}_{pass}_{webhook config UID}`

and if in future we want to expose the order of webhook invocation within one pass, we can append the order to the end:
- `{webhook name}/k8s_{patch|mutation}_{pass}_{webhook config UID}_{order}`

Pros: 
- UID is a 36 character string in current [k8s implementation](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/apimachinery/pkg/util/uuid/uuid.go#L26). We are mostly certain that we won't hit the 63-char limit
- The qualified name segment allows `_`, which is not used by any of the IDs that we are using, makes it easy to parse the name

Cons:
- The prefix is in the webhook name org. Potentially the key name can conflict with a custom annotation (not likely)
- Analyzing the annotations require filtering out the custom annotations, either by doing it manually or through a regex

**Alternatives considered:**
1. We could put less information to the annotation keys and more information to the values, if audit annotations are allowed to be overwritten. Overwriting was designed to be not allowed: https://github.com/kubernetes/kubernetes/pull/58806#discussion_r163921756
2. We can still use `webhook.admission.k8s.io` as `{org}` in the prefix, if we can have an identifier for each webhook which is shorter than the full webhook name (e.g. add a new UID field in the [webhook API](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/api/admissionregistration/v1/types.go#L302-L309), and have the key be `patch.webhook.admission.k8s.io/{webhook UID}_{webhook config UID}`). Problem with this approach is that it's hard to locate a webhook from the webhook UID, and the UID can be fragile to webhook configuration updates.

@jpbetz @sttts WDYT?

Reply:
Another idea: use 
- `{patch|mutation}.webhooks.admission.k8s.io/pass_{pass id}_order_{order id}`

as annotation key, and put {`webhook name`, `configuration name` and `existence or patch`} into the annotation value (better if in a structural format that's easy to parse)

The webhook name was put in the key before because we need something unique for each audit event (each invocation), so that we don't violate the no-overwriting rule. I think a combination of `pass id` + `order id` serves the purpose better.

A disadvantage of this approach is that if someone is analyzing a specific webhook's behavior, they cannot just filter the annotation keys, but need to check the annotation values. The extra effort seems reasonable.

Reply:
Yes, using `webhooks.admission.k8s.io` as the key prefix is what I was expecting.

Having the pass and order field named clearly makes it easy to read the annotations and order them into a chronology of webhook calls so I like how they are presented.

I'm okay with webhook name being in the value.

The only alternative I thought about was doing something like:

key: `*webhooks.admission.k8s.io/_pass_{pass id}`, value: `{webhook_UID_1},{webhook_UID_2}`

But that might be more cumbersome to parse? And I don't know if the value has a size limit.

Reply:
I didn't see a limit for the value size. I want to structure the value as a JSON payload, so we could aggregate all the webhooks into a list. I don't feel strongly either way.